Datapath support for GlobalEgressIPs

sridhargaddam opened the pull request:
This PR implements the datapath support for GlobalEgressIPs
at the namespace level as well as for objects with PodSelectors.

Fixes issue: https://github.com/submariner-io/submariner/issues/1164
Fixes issue: https://github.com/submariner-io/submariner/issues/1165
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

tpantelis commented on pkg/globalnet/controllers/egress_pod_watcher.go:
```diff
@@ -22,14 +22,27 @@ import (
+	}
+
+	var labelSelector string
+	if sel, err := metav1.LabelSelectorAsSelector(podSelector); err == nil {
```
We should report the error.

tpantelis commented on pkg/globalnet/controllers/egress_pod_watcher.go:
```diff
@@ -66,41 +80,60 @@ func (w *podWatcher) arePodsEquivalent(oldObj, newObj *unstructured.Unstructured
+	return nil
+}
+
+func (w *podWatcher) DelPSetEntry(ipAddress string) error {
```
```suggestion
func (w *podWatcher) deleteIPSetEntry(ipAddress string) error {
```

tpantelis commented on pkg/globalnet/controllers/egress_pod_watcher.go:
```diff
@@ -66,41 +80,60 @@ func (w *podWatcher) arePodsEquivalent(oldObj, newObj *unstructured.Unstructured
+		return err
+	}
+
+	return nil
```
```suggestion
	return w.ipSetIface.DelEntry(ipAddress, w.ipSetName)
```

tpantelis commented on pkg/globalnet/controllers/global_egressip_controller.go:
```diff
@@ -260,3 +388,46 @@ func checkStatusChanged(oldStatus, newStatus interface{}, retObj runtime.Object)
+	}
+
+	c.podWatchers[key] = podWatcher
+	c.podWatchers[key].PodSelector = globalEgressIP.Spec.PodSelector
```
```suggestion
	podWatcher.PodSelector = globalEgressIP.Spec.PodSelector
```

tpantelis commented on pkg/globalnet/controllers/global_egressip_controller.go:
```diff
@@ -260,3 +388,46 @@ func checkStatusChanged(oldStatus, newStatus interface{}, retObj runtime.Object)
 }
+
+func isPodSelectorSame(oldSelector, newSelector interface{}) bool {
+	return equality.Semantic.DeepEqual(oldSelector, newSelector)
```
This function is simply a one-liner and has one call site and can be inlined.

tpantelis commented on pkg/ipset/fake/ipset.go:
```diff
@@ -0,0 +1,93 @@
+	"github.com/submariner-io/submariner/pkg/ipset"
+)
+
+type FakeIPSet struct {
```
```suggestion
type IPSet struct {
```

tpantelis commented on pkg/globalnet/controllers/global_egressip_controller.go:
```diff
@@ -113,78 +124,190 @@ func (c *globalEgressIPController) Stop() {
-	if numberOfIPs == nil {
-		one := 1
-		numberOfIPs = &one
+	snatIP := getTargetSNATIPaddress(allocatedIPs)
+	if globalEgressIP.Spec.PodSelector != nil {
+		if err = c.iptIface.AddEgressRulesForPods(key, ipSetName, snatIP, globalNetIPTableMark); err != nil {
+			_ = c.iptIface.RemoveEgressRulesForPods(key, ipSetName, snatIP, globalNetIPTableMark)
+		}
+	} else {
+		if err = c.iptIface.AddEgressRulesForNamespace(key, ipSetName, snatIP, globalNetIPTableMark); err != nil {
+			_ = c.iptIface.RemoveEgressRulesForNamespace(key, ipSetName, snatIP, globalNetIPTableMark)
+		}
```
Can't we call `programGlobalEgressRules` here?

sridhargaddam replied:
yes, I think we can. Thanks @tpantelis

tpantelis commented on pkg/globalnet/controllers/global_egressip_controller.go:
```diff
@@ -194,50 +317,55 @@ func allocateIPs(key string, numberOfIPs *int, pool *ipam.IPPool, status *submar
-		return false
-	}
+		if err := c.ipSetIface.DestroySet(ipSetName); err != nil {
+			klog.Errorf("Error destroying the ipSet %q for %q: %v", ipSetName, key, err)
```
This should really be outside the L355 block b/c on retry it won't re-enter here due to the `podWatcher` removed.

sridhargaddam replied:
Good point @tpantelis

tpantelis commented on pkg/globalnet/controllers/global_egressip_controller.go:
```diff
@@ -113,78 +124,190 @@ func (c *globalEgressIPController) Stop() {
 
-	if err != nil {
-		klog.Errorf("Error stating pod watcher for %q: %v", key, err)
+	ipSetName := c.getIPSetName(key)
+
+	if requeue := c.flushRulesAndReleaseIPs(key, numRequeues, func(allocatedIPs []string) error {
+		if globalEgressIP.Spec.PodSelector != nil {
+			return c.iptIface.RemoveEgressRulesForPods(key, ipSetName,
+				getTargetSNATIPaddress(allocatedIPs), globalNetIPTableMark)
+		} else {
+			return c.iptIface.RemoveEgressRulesForNamespace(key, ipSetName,
+				getTargetSNATIPaddress(allocatedIPs), globalNetIPTableMark)
+		}
+	}, globalEgressIP.Status.AllocatedIPs...); requeue {
```
This could be in a function as it's duplicated with L342.

tpantelis commented on pkg/globalnet/controllers/global_egressip_controller.go:
```diff
@@ -194,50 +317,55 @@ func allocateIPs(key string, numberOfIPs *int, pool *ipam.IPPool, status *submar
 	}
 
-	// TODO - remove IP table rules for previous allocated IPs
+	podWatcher, found := c.podWatchers[key]
+	if found {
+		close(podWatcher.stopCh)
+		delete(c.podWatchers, key)
```
It seems we should stop the `podWatcher` even if `flushRulesAndReleaseIPs` fails above.

tpantelis commented on pkg/globalnet/controllers/types.go:
```diff
@@ -94,11 +99,15 @@ type globalEgressIPController struct {
+	stopCh      chan struct{}
+	ipSetName   string
+	ipSetIface  ipset.Interface
+	PodSelector *metav1.LabelSelector
```
```suggestion
	podSelector *metav1.LabelSelector
```

sridhargaddam reviewed (commented):
Thanks for reviewing @tpantelis

sridhargaddam committed:
Datapath support for GlobalEgressIPs

This PR implements the datapath support for GlobalEgressIPs
at the namespace level as well as for objects with PodSelectors.

Fixes issue: https://github.com/submariner-io/submariner/issues/1164
Fixes issue: https://github.com/submariner-io/submariner/issues/1165
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>
